corehq/apps/hqcase/tasks.py

```diff
@@ -102,33 +102,109 @@ def submit_case(case, new_case_id, domain, source, new_parent_ids=dict()):
-            new_case = copy(case)
+    graph = CaseGraph(cases)
+    for old_case_id in graph.topological_sort():
+        for explosion in range(factor - 1):
+            new_case = copy(cases[old_case_id])
 
             new_case.create = True
-            new_case.case_id = new_case_id_map[old_case_id]
+            new_case.case_id = new_case_ids[old_case_id][explosion]
             new_case.update['cc_exploded_from'] = old_case_id
             new_case.update['cc_explosion_id'] = explosion_id
             new_case.index = {
                 key: IndexAttrs(
-                    i.case_type, new_case_id_map[i.case_id], i.relationship
-                ) for key, i in case.index.iteritems()
+                    i.case_type, new_case_ids[i.case_id][explosion], i.relationship
+                ) for key, i in cases[old_case_id].index.iteritems()
             }
-            new_cases[old_case_id] = ElementTree.tostring(new_case.as_xml())
-        submit_case_blocks(new_cases, domain=domain, device_id='explode_cases_2')
+            queue_case(ElementTree.tostring(new_case.as_xml()), queue)
+
+    if len(queue):
+        submit_case_blocks(queue, domain, user_id=user_id, device_id="explode_cases")
+
+
+class CaseGraph(object):
+    def __init__(self, cases):
+        self.graph = {}         # case indices
+        self.inverse_graph = defaultdict(list)  # reverse indices
+        self.roots = []
+
+        for case_id, case in cases.iteritems():
+            indices = [idx.case_id for idx in case.index.itervalues()]
+            self.graph[case.case_id] = indices
+            for index in indices:
+                self.inverse_graph[index].append(case.case_id)
+
+            if not indices:
+                self.roots.append(case.case_id)
+
+    def get_paths_to_root(self, case_id):
+        """find all paths to the root node
+        """
+        def paths_to_root(case_id, path):
+            path = path + [case_id]
+            if case_id not in self.graph:
+                return [path]
+            if not self.graph[case_id]:  # we made it to the top
+                return [path]
+            paths = []
+            for node in self.graph[case_id]:
+                if node not in path:
+                    newpaths = paths_to_root(node, path)
+                    for newpath in newpaths:
+                        paths.append(newpath)
+            return paths
+
+        return paths_to_root(case_id, [])
+
+    def get_ancestors(self, case_id, include_self=False):
+        paths = self.get_paths_to_root(case_id)
+        ancestors = set()
+        for path in paths:
+            ancestors |= set(path)
+        if not include_self:
+            ancestors.remove(case_id)
+
+        return ancestors
+
+    def topological_sort(self):
+        """returns all cases in topological order
+
+        Using Kahn's algorithm from here:
+        https://en.wikipedia.org/wiki/Topological_sorting
+
+        Assumes there are no cycles in the graph
```

Comment:
I was debating this. This is called with cases that come out of a sync operation, and I don't think we allow you to have cycles there. I've added a check for now, but I'm not sure it will ever catch anything, and has to loop through each of the cases again. What do you think?

Reply:
I think it's a good conservative move, even if it's unlikely to ever catch anything. The run time expense is likely not prohibitive, and it will be nice to know that it fails predictably if it finds unexpected cycles (rather than infinite recursion or loop forever or whatever).